api/pkg/s3/objectget.go

```diff
@@ -24,65 +33,152 @@ import (
 		return
 	}
+	object := objMetaRes.Object
 
-	object.ObjectKey = objectKey
-	object.BucketName = bucketName
-	var client datastore.DataStoreAdapter
-	if backendname != "" {
-		client = getBackendByName(ctx, s, backendname)
-	} else {
-		client = getBackendClient(ctx, s, bucketName)
+	// Get request range.
+	var hrange *HttpRange
+	rangeHeader := request.HeaderParameter("Range")
+	if rangeHeader != "" {
+		if hrange, err = ParseRequestRange(rangeHeader, object.Size); err != nil {
+			// Handle only ErrorInvalidRange
+			// Ignore other parse error and treat it as regular Get request like Amazon S3.
+			if err == ErrorInvalidRange {
+				WriteErrorResponse(response, request, s3error.ErrInvalidRange)
+				return
+			}
+			// log the error.
+			log.Errorln("invalid request range, err:", err)
+		}
 	}
-	if client == nil {
-		response.WriteError(http.StatusInternalServerError, NoSuchBackend.Error())
+
+	// Validate pre-conditions if any.
+	if err = checkPreconditions(request.Request.Header, object); err != nil {
+		// set object-related metadata headers
+		response.AddHeader("Last-Modified", time.Unix(object.LastModified, 0).UTC().Format(http.TimeFormat))
+
+		if object.Etag != "" {
+			response.ResponseWriter.Header()["ETag"] = []string{"\"" + object.Etag + "\""}
+		}
+		if err == s3error.ContentNotModified { // write only header if is a 304
+			WriteErrorResponseHeaders(response, err)
+		} else {
+			WriteErrorResponse(response, request, err)
+		}
 		return
 	}
-	log.Infof("enter the download method")
-	body, s3err := client.GET(&object, ctx, int64(start), int64(end))
-	log.Infof("out  the download method")
-	if s3err != NoError {
-		response.WriteError(http.StatusInternalServerError, s3err.Error())
+
+	// Get the object.
+	startOffset := int64(0)
+	length := object.Size
+	if hrange != nil {
+		startOffset = hrange.OffsetBegin
+		length = hrange.GetLength()
+	}
+	stream, err := s.s3Client.GetObject(ctx, &pb.GetObjectInput{Bucket:bucketName, Key:objectKey, Offset: startOffset, Length: length,})
+	if err != nil {
+		log.Errorln("get object failed, err:%v", err)
+		WriteErrorResponse(response, request, err)
 		return
 	}
-	buf := new(bytes.Buffer)
-	buf.ReadFrom(body)
-	response.Write(buf.Bytes())
-	*/
-	log.Infof("Init multipart upload[bucketName=%s, objectKey=%s] successfully.\n", bucketName, objectKey)
+	defer stream.Close()
+
+	// Indicates if any data was written to the http.ResponseWriter
+	dataWritten := false
+	// io.Writer type which keeps track if any data was written.
+	writer := funcToWriter(func(p []byte) (int, error) {
+		if !dataWritten {
+			// Set headers on the first write.
+			// Set standard object headers.
+			SetObjectHeaders(response, object, hrange)
+
+			// Set any additional requested response headers.
+			setGetRespHeaders(response.ResponseWriter, request.Request.URL.Query())
+			dataWritten = true
+		}
+		n, err := response.Write(p)
+		if n > 0 {
+
+		}
+		return n, err
+	})
+
+	s3err := int32(0)
```

Comment:
Should the initial value of s3err be ErrNoErr?

Reply:
thanks, I'll fix it